{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Basic Analysis and Visualization of the Spatial GIFT-seq dataset",
   "id": "15a2f490208da761"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "import gc\n",
    "import os\n",
    "# os.chdir(\"/data1/lareauc/users/varelaa/giftwrap/notebooks\")\n",
    "os.chdir(\"/home/varelaa/quick_analysis/\")\n",
    "import sys\n",
    "sys.path.append(\"./\")\n",
    "from figures import *\n",
    "\n",
    "import anndata as ad\n",
    "import giftwrap as gw\n",
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import spatialdata as sd\n",
    "import spatialdata_io as sio\n",
    "import matplotlib as mpl\n",
    "from scipy.stats import gaussian_kde\n",
    "from shapely import Polygon\n",
    "from geopandas import GeoDataFrame\n",
    "from shapely import Polygon\n",
    "from spatialdata import polygon_query\n",
    "from spatialdata.models import ShapesModel\n",
    "from spatialdata.transformations import Identity\n",
    "#!pip install adjustText\n",
    "import adjustText\n",
    "mpl.rcParams['figure.dpi'] = 300\n",
    "RESOLUTION=16  # in um"
   ],
   "id": "fca735f2ef8ca6f8"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Load our datasets",
   "id": "e05cabf4fb6ce7ba"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# # First, our GIFT-seq dataset\n",
    "# dual_probe_gdata = gw.read_h5_file(\"/data1/lareauc/projects/gapfill/analysis/20250708_visiumHD_benchmark/DualPanel_GIFTWRAP_WTA_FILTER/counts.1.h5\")\n",
    "# gapfill_gdata = gw.read_h5_file(\"/data1/lareauc/projects/gapfill/analysis/20250708_visiumHD_benchmark/GapFill_GIFTWRAP_WITH_WTA_TEST_UPDATE/counts.1.h5\")\n",
    "#\n",
    "# # Next the WTAs\n",
    "# dual_probe_wta = sio.visium_hd(\n",
    "#     \"/data1/lareauc/projects/gapfill/analysis/20250708_visiumHD_benchmark/CL_DualPanel_WTA/outs\",\n",
    "#      dataset_id=''\n",
    "# )\n",
    "# gapfill_wta = sio.visium_hd(\n",
    "#     \"/data1/lareauc/projects/gapfill/analysis/20250708_visiumHD_benchmark/CL_GapFill_WTA/outs\",\n",
    "#      dataset_id=''\n",
    "# )\n",
    "\n",
    "# First, our GIFT-seq dataset\n",
    "dual_probe_gdata = gw.read_h5_file(\"20250708_visiumHD_benchmark/DualPanel_GIFTWRAP_WTA_FILTER/counts.1.h5\")\n",
    "# gapfill_gdata = gw.read_h5_file(\"20250708_visiumHD_benchmark/GapFill_GIFTWRAP_WITH_WTA_TEST_UPDATE/counts.1.h5\")\n",
    "gapfill_gdata = gw.read_h5_file(\"20250708_visiumHD_benchmark/GapFill_GIFTWRAP_WITH_WTA_TEST_UPDATE_IGO/counts.1.h5\")\n",
    "# dual_probe_gdata= gw.read_h5_file(\"20250708_visiumHD_benchmark/GapFill_GIFTWRAP_WITH_WTA_TEST_UPDATE_IGO/counts.1.h5\")\n",
    "\n",
    "# Next the WTAs\n",
    "dual_probe_wta = sio.visium_hd(\n",
    "    \"20250708_visiumHD_benchmark/CL_DualPanel_WTA/outs\",\n",
    "     dataset_id=''\n",
    ")\n",
    "# Convert to ZARR\n",
    "dual_probe_wta.write(\"./dual_probe_wta.zarr\", overwrite=True)\n",
    "# Re-read\n",
    "del dual_probe_wta\n",
    "gc.collect()\n",
    "dual_probe_wta = sd.read_zarr(\"./dual_probe_wta.zarr\")\n",
    "gapfill_wta = sio.visium_hd(\n",
    "    \"20250708_visiumHD_benchmark/CL_GapFill_WTA/outs\",\n",
    "     dataset_id=''\n",
    ")\n",
    "# Convert to ZARR\n",
    "gapfill_wta.write(\"./gapfill_wta.zarr\", overwrite=True)\n",
    "# Re-read\n",
    "del gapfill_wta\n",
    "gc.collect()\n",
    "gapfill_wta = sd.read_zarr(\"./gapfill_wta.zarr\")\n",
    "\n",
    "# gapfill_gdata = gw.read_h5_file(\"/data1/lareauc/projects/gapfill/analysis/20250226_visiumHD/20250224_M81S_GF_CL_GIFTWRAP_REDUX/counts.1.h5\")\n",
    "# gapfill_wta = sio.visium_hd(\n",
    "#     \"/data1/lareauc/projects/gapfill/analysis/20250226_visiumHD/Visium_CL_WTA/outs\",\n",
    "#      dataset_id=''\n",
    "# )\n",
    "\n",
    "# From: https://github.com/clareaulab/gift_reproducibility/blob/main/figure_CL_dual/data/3cl_predicted_genotypes.csv\n",
    "celltype_genotypes_df = pd.read_csv(\"cell_line_data.csv\", index_col=0)\n",
    "celltype_genotypes_df"
   ],
   "id": "909ebfa069c9b978"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "dual_probe_gdata.X = dual_probe_gdata.layers['X_pcr_threshold_5']\n",
    "gapfill_gdata.X = gapfill_gdata.layers['X_pcr_threshold_5']"
   ],
   "id": "a9132166fbde1e4c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Map genotypes\n",
    "annotated_genotypes = celltype_genotypes_df.name.unique().tolist()\n",
    "wt_alleles = dict()\n",
    "alt_alleles = dict()\n",
    "celltype_genotypes = {\n",
    "    \"HEL\": dict(),\n",
    "    \"K562\": dict(),\n",
    "    \"SET2\": dict(),\n",
    "}\n",
    "celltype_annotated = {\n",
    "    \"HEL\": dict(),\n",
    "    \"K562\": dict(),\n",
    "    \"SET2\": dict()\n",
    "}\n",
    "for i, row in celltype_genotypes_df.iterrows():\n",
    "    wt_alleles[row['name']] = row[\"gapfill_from_transcriptome\"]\n",
    "    alt_alleles[row['name']] = row[\"gap_probe_sequence\"]\n",
    "    if row['genotype_from_bulk'] == 'heterozygous':\n",
    "        celltype_annotated[row['cell_type']][row[\"name\"]] = \"HET\"\n",
    "    elif row['genotype_from_bulk'] == 'homozygous_ref':\n",
    "        celltype_annotated[row['cell_type']][row[\"name\"]] = \"REF\"\n",
    "    elif row['genotype_from_bulk'] == 'homozygous_alt':\n",
    "        celltype_annotated[row['cell_type']][row[\"name\"]] = \"ALT\"\n",
    "    first_gf = row[\"0\"]\n",
    "    second_gf = row[\"1\"]\n",
    "    celltype_genotypes[row['cell_type']][row[\"name\"]] = [first_gf]\n",
    "    if not pd.isna(second_gf) and second_gf != \"\":\n",
    "        celltype_genotypes[row['cell_type']][row[\"name\"]].append(second_gf)"
   ],
   "id": "3d4ec70824b402e6"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Manually create polygons for cell lines\n",
    "def numpy_to_shapely(poly):\n",
    "    return Polygon([(x, y) for x, y in poly])\n",
    "\n",
    "dual_probe_polygons = {\n",
    "    \"HEL\": numpy_to_shapely(np.array([[9_000, 43_000], [35_000, 45_000], [35_000, 57_000], [9_000, 57_000]])),\n",
    "    \"K562\": numpy_to_shapely(np.array([[9_000, 58_000], [22_000, 58_000], [22_000, 75_000], [9_000, 75_000]])),\n",
    "    \"SET2\": numpy_to_shapely(np.array([[23_000, 58_000], [40_000, 58_000], [40_000, 75_000], [23_000, 75_000]])),\n",
    "}\n",
    "\n",
    "dual_probe_wta.tables['square_002um'].obs['cell_line'] = 'N/A'\n",
    "dual_probe_wta.tables['square_008um'].obs['cell_line'] = 'N/A'\n",
    "dual_probe_wta.tables['square_016um'].obs['cell_line'] = 'N/A'\n",
    "for cell_line, poly in dual_probe_polygons.items():\n",
    "    filtered = polygon_query(\n",
    "        dual_probe_wta,\n",
    "        polygon=poly,\n",
    "        target_coordinate_system=\"\",\n",
    "    )\n",
    "    dual_probe_wta.tables['square_002um'].obs.loc[filtered.tables['square_002um'].obs_names, 'cell_line'] = cell_line\n",
    "    dual_probe_wta.tables['square_008um'].obs.loc[filtered.tables['square_008um'].obs_names, 'cell_line'] = cell_line\n",
    "    dual_probe_wta.tables['square_016um'].obs.loc[filtered.tables['square_016um'].obs_names, 'cell_line'] = cell_line\n",
    "\n",
    "gapfill_polygons = {\n",
    "    # Reordered last two vertices\n",
    "    \"HEL\": numpy_to_shapely(np.array([[10_000, 12_000], [40_000, 12_000], [40_000, 28_000], [10_000, 28_000]])),\n",
    "    \"K562\": numpy_to_shapely(np.array([\n",
    "        [10_000, 27_000], [19_000, 27_000], [19_000, 37_000],\n",
    "        [19_500, 38_000], [19_000, 40_000], [10_000, 42_000],\n",
    "        [10_000, 40_000]\n",
    "    ])),\n",
    "    \"SET2\": numpy_to_shapely(np.array([\n",
    "        [33_500, 26_500], [19_500, 37_000], [19_000, 40_000],\n",
    "        [18_000, 43_000], [33_500, 44_500]\n",
    "    ])),\n",
    "}\n",
    "\n",
    "gapfill_wta.tables['square_002um'].obs['cell_line'] = 'N/A'\n",
    "gapfill_wta.tables['square_008um'].obs['cell_line'] = 'N/A'\n",
    "gapfill_wta.tables['square_016um'].obs['cell_line'] = 'N/A'\n",
    "for cell_line, poly in gapfill_polygons.items():\n",
    "    filtered = polygon_query(\n",
    "        gapfill_wta,\n",
    "        polygon=poly,\n",
    "        target_coordinate_system=\"\",\n",
    "    )\n",
    "    gapfill_wta.tables['square_002um'].obs.loc[filtered.tables['square_002um'].obs_names, 'cell_line'] = cell_line\n",
    "    gapfill_wta.tables['square_008um'].obs.loc[filtered.tables['square_008um'].obs_names, 'cell_line'] = cell_line\n",
    "    gapfill_wta.tables['square_016um'].obs.loc[filtered.tables['square_016um'].obs_names, 'cell_line'] = cell_line\n",
    "\n",
    "#\n",
    "# gapfill_polygons = {\n",
    "#     \"HEL\": numpy_to_shapely(np.array([[5_000, 5_000], [35_000, 5_000], [35_000, 19_000], [5_000, 19_000]])),\n",
    "#     \"K562\": numpy_to_shapely(np.array([[5_000, 20_000], [22_000, 20_000], [22_000, 40_000], [5_000, 40_000]])),\n",
    "#     \"SET2\": numpy_to_shapely(np.array([[18_000, 19_000], [40_000, 19_000], [40_000, 40_000], [18_000, 40_000]])),\n",
    "# }\n",
    "#\n",
    "# gapfill_wta.tables['square_002um'].obs['cell_line'] = 'N/A'\n",
    "# gapfill_wta.tables['square_008um'].obs['cell_line'] = 'N/A'\n",
    "# gapfill_wta.tables['square_016um'].obs['cell_line'] = 'N/A'\n",
    "# for cell_line, poly in gapfill_polygons.items():\n",
    "#     filtered = polygon_query(\n",
    "#         gapfill_wta,\n",
    "#         polygon=poly,\n",
    "#         target_coordinate_system=\"\",\n",
    "#     )\n",
    "#     gapfill_wta.tables['square_002um'].obs.loc[filtered.tables['square_002um'].obs_names, 'cell_line'] = cell_line\n",
    "#     gapfill_wta.tables['square_008um'].obs.loc[filtered.tables['square_008um'].obs_names, 'cell_line'] = cell_line\n",
    "#     gapfill_wta.tables['square_016um'].obs.loc[filtered.tables['square_016um'].obs_names, 'cell_line'] = cell_line"
   ],
   "id": "2fee58c7fdbb3755"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Crop to tissue area\n",
    "dual_probe_wta = dual_probe_wta.query.bounding_box(\n",
    "    axes=(\"x\", \"y\"),\n",
    "    min_coordinate=np.array([0, 30_000]),\n",
    "    max_coordinate=np.array([40_000, 80_000]),\n",
    "    target_coordinate_system=\"\",\n",
    "    filter_table=False\n",
    ")\n",
    "dual_probe_wta"
   ],
   "id": "378a4e6fdb545afa"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Crop to tissue area\n",
    "gapfill_wta = gapfill_wta.query.bounding_box(\n",
    "    axes=(\"x\", \"y\"),\n",
    "    min_coordinate=np.array([0, 0]),\n",
    "    max_coordinate=np.array([40_000, 45_000]),\n",
    "    target_coordinate_system=\"\",\n",
    "    filter_table=False\n",
    ")\n",
    "gapfill_wta"
   ],
   "id": "d320ae678f2d277f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_HE(dual_probe_wta)",
   "id": "8cb6314752109132"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_HE(gapfill_wta)",
   "id": "a9ed0e7be297936c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Plot the tissue with the annotations\n",
    "ax = dual_probe_wta.pl.render_images(f\"_hires_image\", alpha=0.8) \\\n",
    "    .pl.render_shapes(element=f'_square_{RESOLUTION:03d}um', color='cell_line', na_color=None) \\\n",
    "    .pl.show(coordinate_systems=\"\", figsize=(25, 25), na_in_legend=False, title=\"Dual Probe Selection\", return_ax=True)\n",
    "\n",
    "# Remove the x and y ticks, tick labels\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "# Rename the x axis and y axis\n",
    "ax.set_xlabel(\"Spatial 1\")\n",
    "ax.set_ylabel(\"Spatial 2\")\n",
    "\n",
    "ax"
   ],
   "id": "1a0c642adafe690f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Plot the tissue with the annotations\n",
    "ax = gapfill_wta.pl.render_images(f\"_hires_image\", alpha=0.8) \\\n",
    "    .pl.render_shapes(element=f'_square_{RESOLUTION:03d}um', color='cell_line', na_color=None) \\\n",
    "    .pl.show(coordinate_systems=\"\", figsize=(25, 25), na_in_legend=False, title=\"GapFill Probe Selection\", return_ax=True)\n",
    "\n",
    "# Remove the x and y ticks, tick labels\n",
    "ax.set_xticks([])\n",
    "ax.set_yticks([])\n",
    "# Rename the x axis and y axis\n",
    "ax.set_xlabel(\"Spatial 1\")\n",
    "ax.set_ylabel(\"Spatial 2\")\n",
    "\n",
    "ax"
   ],
   "id": "258c91cca2770c20"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "dual_probe_gdata",
   "id": "d92375424361548d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "gapfill_gdata",
   "id": "fa7eb5354d70ef32"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "dual_probe_wta",
   "id": "bacb3d77b0c6d6b2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "gapfill_wta",
   "id": "d7f2fcfe8a86744d"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Genotyping\n",
    "\n",
    "We will do basic preprocessing and genotyping calls on the GIFT-seq datasets"
   ],
   "id": "10ea7aeb793945ce"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## First, the dual-probe dataset",
   "id": "1eda22ed8e76319b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# We don't expect any gapfills, so we should filter out UMIs with a gapfill\n",
    "probes_with_gapfill = dual_probe_gdata.var[dual_probe_gdata.var['gapfill'] != ''].index\n",
    "# Count number of probes and the number of umis this affects\n",
    "print(f\"Number of probes with gapfill: {len(probes_with_gapfill)} / {dual_probe_gdata.var.shape[0]}\")\n",
    "umis_with_gapfill = np.sum(dual_probe_gdata[:, probes_with_gapfill].X)\n",
    "print(f\"Number of UMIs with gapfill: {umis_with_gapfill} / {dual_probe_gdata.X.sum()}\")\n",
    "# Drop them\n",
    "dual_probe_gdata = dual_probe_gdata[:, dual_probe_gdata.var['gapfill'] == '']\n",
    "dual_probe_gdata"
   ],
   "id": "a1bc4853699c58a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "# dual_probe_gdata.X = dual_probe_gdata.layers['X_pcr_threshold_5']",
   "id": "fcd8be97562a20c3"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Since the dual-probe dataset isn't formatted correctly for genotyping, we will recreate the AnnData to look more like a vanilla GIFT-seq dataset",
   "id": "2a82c876e5cc0e2b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Select all the 0bp control probes, these will not be used for genotyping\n",
    "zerobp_probes = dual_probe_gdata.var[dual_probe_gdata.var.probe.str.contains(\"0bp\")].probe.values\n",
    "mut_wt_pairs = []\n",
    "# Now for each non 0bp probe, we will find the mutant probe and its associated WT probe\n",
    "for probe in dual_probe_gdata.var.probe.values:\n",
    "    if \"0bp\" in probe:\n",
    "        continue\n",
    "    if \">\" in probe:\n",
    "        orig = probe.split(\">\")[0]\n",
    "        alt = probe.split(\">\")[1]\n",
    "        if orig[-1] == alt:  # This was a WT probe\n",
    "            continue\n",
    "        wt_probe = f\"{orig}>{orig[-1]}\"\n",
    "        mut_wt_pairs.append((probe, wt_probe, alt, orig[-1]))\n",
    "    else:\n",
    "        print(f\"Unexpected probe name: {probe}\")\n",
    "\n",
    "# Manually add the probes that were not named correctly\n",
    "mut_wt_pairs.append((\"BCR-ABL c.fusion\", \"BCR-ABL null\", 'fusion', 'null'))\n",
    "mut_wt_pairs.append((\"TP53 c.405insC\", \"TP53 c.405\", \"insC\", \"ref\"))\n",
    "\n",
    "print(f\"Dropping the following probes that could not be paired: {set(dual_probe_gdata.var.probe) - set(zerobp_probes) - set([a for a, b,c,d in mut_wt_pairs]) - set([b for a, b,c,d in mut_wt_pairs])}\")\n",
    "# Create a new AnnData with just the probes we want\n",
    "new_var = dict(\n",
    "    probe=[],\n",
    "    gene=[],\n",
    "    gapfill=[],\n",
    ")\n",
    "new_X = np.zeros((dual_probe_gdata.n_obs, 2 * len(mut_wt_pairs) + zerobp_probes.shape[0]), dtype=np.int32)\n",
    "new_layers = {k: np.zeros((dual_probe_gdata.n_obs, 2 * len(mut_wt_pairs) + zerobp_probes.shape[0]), dtype=np.int32) for k in dual_probe_gdata.layers.keys() if 'X' in k and int(k.split(\"_\")[-1]) < 25}\n",
    "for i, probe in enumerate(zerobp_probes):\n",
    "    idx = dual_probe_gdata.var.index[dual_probe_gdata.var.probe == probe][0]\n",
    "    new_var['probe'].append(probe)\n",
    "    new_var['gene'].append(dual_probe_gdata.var.loc[idx, 'gene'])\n",
    "    new_var['gapfill'].append(dual_probe_gdata.var.loc[idx, 'gapfill'])\n",
    "    new_X[:, i] = dual_probe_gdata[:, idx].X.toarray().flatten()\n",
    "    for k in list(new_layers.keys()):\n",
    "        new_layers[k][:, i] = dual_probe_gdata[:, idx].layers[k].toarray().flatten()\n",
    "for j, (mut, wt, mut_genotype, wt_genotype) in enumerate(mut_wt_pairs):\n",
    "    wt_idx = dual_probe_gdata.var.index[dual_probe_gdata.var.probe == wt][0]\n",
    "    mut_idx = dual_probe_gdata.var.index[dual_probe_gdata.var.probe == mut][0]\n",
    "    new_var['probe'].append(mut)\n",
    "    new_var['gene'].append(dual_probe_gdata.var.loc[wt_idx, 'gene'])\n",
    "    new_var['gapfill'].append(wt_genotype)\n",
    "    new_X[:, len(zerobp_probes) + 2 * j] = dual_probe_gdata[:, wt_idx].X.toarray().flatten()\n",
    "    new_var['probe'].append(mut)\n",
    "    new_var['gene'].append(dual_probe_gdata.var.loc[wt_idx, 'gene'])\n",
    "    new_var['gapfill'].append(mut_genotype)\n",
    "    new_X[:, len(zerobp_probes) + 2 * j + 1] = dual_probe_gdata[:, mut_idx].X.toarray().flatten()\n",
    "    for k in list(new_layers.keys()):\n",
    "        new_layers[k][:, len(zerobp_probes) + 2 * j] = dual_probe_gdata[:, wt_idx].layers[k].toarray().flatten()\n",
    "        new_layers[k][:, len(zerobp_probes) + 2 * j + 1] = dual_probe_gdata[:, mut_idx].layers[k].toarray().flatten()\n",
    "\n",
    "var_df = pd.DataFrame(new_var)\n",
    "var_df['probe_gapfill'] = var_df['probe'] + \"_\" + var_df['gapfill']\n",
    "var_df = var_df.set_index('probe_gapfill')\n",
    "dual_probe_gdata = ad.AnnData(\n",
    "    X=new_X,\n",
    "    obs=dual_probe_gdata.obs.copy(),\n",
    "    var=var_df,\n",
    "    uns=dual_probe_gdata.uns.copy(),\n",
    "    layers=new_layers\n",
    ")\n",
    "dual_probe_gdata"
   ],
   "id": "cb3ab78f6522f3b6"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Genotype\n",
    "dual_probe_gdata = gw.tl.call_genotypes(\n",
    "    dual_probe_gdata\n",
    ")\n",
    "dual_probe_gdata"
   ],
   "id": "f7b4690d55a118f0"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Join with the spatialdata\n",
    "dual_probe_sdata = gw.sp.join_with_wta(dual_probe_wta, dual_probe_gdata)\n",
    "dual_probe_sdata"
   ],
   "id": "61a25b4613598a3f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print_summary_stats(dual_probe_sdata, RESOLUTION, include_0bp=True)",
   "id": "9c21230f9a7904aa"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print_summary_stats(dual_probe_sdata, RESOLUTION, include_0bp=False)",
   "id": "1dff54ddff1bd0d"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Now, compute the genotyping efficiency metrics",
   "id": "18043e78247887a5"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Plot the WTA library size\n",
    "plot_library_size(dual_probe_sdata, table='', resolution=RESOLUTION)"
   ],
   "id": "7cbe801142d5b985"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Plot the GIFT-seq library size\n",
    "plot_library_size(dual_probe_sdata, table='gf', resolution=RESOLUTION)"
   ],
   "id": "980662ecf43f76ef"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_library_size(dual_probe_sdata, table='gf', resolution=RESOLUTION, include_0bp=True)",
   "id": "5d8ae9ea2bff21d5"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_sites_genotyped(dual_probe_sdata, resolution=RESOLUTION)",
   "id": "b5c6701358e9a51d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_sites_genotyped(dual_probe_sdata, resolution=RESOLUTION, at_least_one=True)",
   "id": "f214e4eed7c56556"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "compare_library_size_per_bin(dual_probe_sdata, resolution=RESOLUTION)",
   "id": "4d19dde84dfd7d2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "compare_library_size_per_bin(dual_probe_sdata, resolution=RESOLUTION, include_0bp=True)",
   "id": "1e82d207630195ad"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_relative_efficiency(dual_probe_sdata, resolution=RESOLUTION, min_0bp_count=30_000, min_gf_count=20_000)",
   "id": "883137dc83f49782"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "plot_genotype_umi_comparison(dual_probe_sdata, 'K562', 'HEL',\n",
    "    annotated_genotypes,\n",
    "    celltype_genotypes,\n",
    "    wt_alleles,\n",
    "    alt_alleles,\n",
    "    RESOLUTION\n",
    ")"
   ],
   "id": "e84a026362dc68f9"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "plot_genotype_umi_comparison(dual_probe_sdata, 'K562', 'SET2',\n",
    "    annotated_genotypes,\n",
    "    celltype_genotypes,\n",
    "    wt_alleles,\n",
    "    alt_alleles,\n",
    "    RESOLUTION\n",
    ")"
   ],
   "id": "a21c99dd6b2fbe83"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "plot_genotype_umi_comparison(dual_probe_sdata, 'SET2', 'HEL',\n",
    "    annotated_genotypes,\n",
    "    celltype_genotypes,\n",
    "    wt_alleles,\n",
    "    alt_alleles,\n",
    "    RESOLUTION\n",
    ")"
   ],
   "id": "b9d6ff3fc516b0ac"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "pseudobulk_genotype_table(dual_probe_sdata, 'JAK2 c.1849G>T', \"G\", \"T\", {\n",
    "    \"K562\": \"WT\",\n",
    "    \"HEL\": \"ALT\",\n",
    "    \"SET2\": \"HET\"\n",
    "})"
   ],
   "id": "3bac08f439dd3eb6"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# dual_probe_sdata.tables['gf_square_002um'].X = dual_probe_sdata.tables['gf_square_002um'].layers['X_pcr_threshold_5']  # Adjust PCR threshold\n",
    "# dual_probe_sdata.tables['gf_square_008um'].X = dual_probe_sdata.tables['gf_square_008um'].layers['X_pcr_threshold_5']\n",
    "# dual_probe_sdata.tables['gf_square_016um'].X = dual_probe_sdata.tables['gf_square_016um'].layers['X_pcr_threshold_5']"
   ],
   "id": "32e9327105a6dbe7"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "genotype_psuedobulk_accuracy_by_pcr(dual_probe_sdata, 'JAK2 c.1849G>T', \"G\", \"T\", {\n",
    "    \"K562\": \"WT\",\n",
    "    \"HEL\": \"ALT\",\n",
    "    \"SET2\": \"HET\"\n",
    "}, ['K562', 'HEL'], max_threshold=25)"
   ],
   "id": "7db75d955c4caa3a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_library_specific_probe(dual_probe_sdata, 'JAK2 c.1849G>T', \"G\", RESOLUTION)",
   "id": "22a8b9dd639b1589"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_library_specific_probe(dual_probe_sdata, 'JAK2 c.1849G>T', \"T\", RESOLUTION)",
   "id": "fc71b08d7d14e6dd"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "genotype_cell_line_barplots(dual_probe_sdata, 'JAK2 c.1849G>T', \"G\", \"T\", RESOLUTION)",
   "id": "fb7eaccec4c88752"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "genotype_accuracy_barplot(dual_probe_sdata, 'JAK2 c.1849G>T', \"G\", \"T\", {\n",
    "    \"K562\": \"WT\",\n",
    "    \"HEL\": \"ALT\",\n",
    "    \"SET2\": \"HET\"\n",
    "}, RESOLUTION)"
   ],
   "id": "504197fedfbc0c2d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "genotype_accuracy_barplot(dual_probe_sdata, 'JAK2 c.1849G>T', \"G\", \"T\", {\n",
    "    \"K562\": \"WT\",\n",
    "    \"HEL\": \"ALT\",\n",
    "    \"SET2\": \"HET\"\n",
    "}, RESOLUTION, filter_na=False)"
   ],
   "id": "3f24a8c729260978"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "for probe in dual_probe_gdata.var.probe.unique():\n",
    "    if probe not in annotated_genotypes:\n",
    "        continue\n",
    "    wt_allele = \"\" if \">\" not in probe else probe.split(\">\")[0][-1]\n",
    "    alt_allele = \"\" if \">\" not in probe else probe.split(\">\")[1]\n",
    "    ct_dict = {\n",
    "        \"HEL\": \"HET\" if len(celltype_genotypes[\"HEL\"][probe]) > 1 else (\"WT\" if wt_alleles[probe] in celltype_genotypes[\"HEL\"][probe] else \"ALT\"),\n",
    "        \"K562\": \"HET\" if len(celltype_genotypes[\"K562\"][probe]) > 1 else (\"WT\" if wt_alleles[probe] in celltype_genotypes[\"K562\"][probe] else \"ALT\"),\n",
    "        \"SET2\": \"HET\" if len(celltype_genotypes[\"SET2\"][probe]) > 1 else (\"WT\" if wt_alleles[probe] in celltype_genotypes[\"SET2\"][probe] else \"ALT\"),\n",
    "    }\n",
    "    fig, axes = plt.subplots(1, 2, figsize=(8, 4))\n",
    "    genotype_cell_line_barplots(dual_probe_sdata, probe, wt_allele, alt_allele, RESOLUTION, ax=axes[0])\n",
    "    genotype_accuracy_barplot(dual_probe_sdata, probe, wt_allele, alt_allele, ct_dict, RESOLUTION, filter_na=True, ax=axes[1])\n",
    "    plt.suptitle(probe)\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "    plt.clf()"
   ],
   "id": "4427e0e5058a28d6"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Now some basic visualization and analysis",
   "id": "e477a6808aaa80a5"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Sort probes by capture rate and plot each on the spatial coords\n",
    "probes = dual_probe_gdata.var.probe[~dual_probe_gdata.var.probe.str.contains(\"0bp\")]\n",
    "probes = sorted(probes, key=lambda x: dual_probe_gdata[:, dual_probe_gdata.var.probe == x].X.sum(), reverse=True)\n",
    "\n",
    "for probe in probes:\n",
    "    print(f\"Plotting {probe}\")\n",
    "    plot_genotypes(dual_probe_sdata, probe, resolution=RESOLUTION)\n",
    "    plt.show()\n",
    "    plt.clf()"
   ],
   "id": "99a578e4ea4c9d9c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Cluster\n",
    "dual_probe_sdata = gw.sp.recipe_spatial_expression_coclustering(\n",
    "    dual_probe_sdata,\n",
    "    table_name=f\"square_{RESOLUTION:03d}um\",\n",
    "    n_highly_variable_genes=1000\n",
    ")\n",
    "dual_probe_sdata"
   ],
   "id": "25e278770837dcb"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Impute genotypes\n",
    "dual_probe_sdata = gw.sp.impute_genotypes(\n",
    "    dual_probe_sdata,\n",
    "    cluster_key='spatio_expression_coclustering',\n",
    "    resolution=f'square_{RESOLUTION:03d}um',\n",
    "    hold_out=0.05\n",
    ")\n",
    "dual_probe_sdata"
   ],
   "id": "4096dbc117f70f3b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Plot imputed genotypes\n",
    "for probe in probes:\n",
    "    print(f\"Plotting {probe}\")\n",
    "    plot_genotypes(dual_probe_sdata, probe, resolution=RESOLUTION, imputed=True)\n",
    "    plt.show()\n",
    "    plt.clf()"
   ],
   "id": "ef89a54edac472dc"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Now, the GapFill Dataset",
   "id": "53f2057029f7cb48"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## First, we will do some basic preprocessing",
   "id": "63bd1e80c13d6bc1"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "gw.pp.filter_gapfills(gapfill_gdata, min_cells=10)\n",
    "gapfill_gdata = gw.tl.call_genotypes(\n",
    "    gapfill_gdata\n",
    ")\n",
    "# gapfill_gdata.X = gapfill_gdata.X - gapfill_gdata.layers['X_pcr_threshold_5']  To isolate filtered out umis\n",
    "gapfill_gdata"
   ],
   "id": "23b2e8f84129813b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "gapfill_gdata.obsm['genotype']['JAK2 c.1849G>T'].value_counts()",
   "id": "bf6e32995f6db1b4"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "gapfill_sdata = gw.sp.join_with_wta(gapfill_wta, gapfill_gdata)\n",
    "gapfill_sdata"
   ],
   "id": "9b08a172cdff3578"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print_summary_stats(gapfill_sdata, RESOLUTION, include_0bp=True)",
   "id": "1b42099391ae448c"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "print_summary_stats(gapfill_sdata, RESOLUTION, include_0bp=False)",
   "id": "9c4aa090146386da"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Now, compute the genotyping efficiency metrics",
   "id": "42b7e948c4485d37"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Plot the WTA library size\n",
    "plot_library_size(gapfill_sdata, table='', resolution=RESOLUTION)"
   ],
   "id": "e0c6c9e9b9933281"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Plot the GIFT-seq library size\n",
    "plot_library_size(gapfill_sdata, table='gf', resolution=RESOLUTION)"
   ],
   "id": "cb673f6d5ada9de9"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_library_size(gapfill_sdata, table='gf', resolution=RESOLUTION, include_0bp=True)",
   "id": "ba1aa75dcde8545b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_sites_genotyped(gapfill_sdata, resolution=RESOLUTION)",
   "id": "3a785b912f2e456b"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_sites_genotyped(gapfill_sdata, resolution=RESOLUTION, at_least_one=True)",
   "id": "8c5a060648e2bdc8"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "compare_library_size_per_bin(gapfill_sdata, resolution=RESOLUTION)",
   "id": "9798819e64431c72"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "compare_library_size_per_bin(gapfill_sdata, resolution=RESOLUTION, include_0bp=True)",
   "id": "ee34b70a84a87c5d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "plot_relative_efficiency(gapfill_sdata, resolution=RESOLUTION, min_0bp_count=30_000, min_gf_count=20_000)",
   "id": "c1670e512ed647fc"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "plot_genotype_umi_comparison(gapfill_sdata, 'K562', 'HEL',\n",
    "                             annotated_genotypes,\n",
    "                             celltype_genotypes,\n",
    "                             wt_alleles,\n",
    "                             alt_alleles,\n",
    "                             RESOLUTION\n",
    "                             )"
   ],
   "id": "cb9ece0d212ac94"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "plot_genotype_umi_comparison(gapfill_sdata, 'K562', 'SET2',\n",
    "                             annotated_genotypes,\n",
    "                             celltype_genotypes,\n",
    "                             wt_alleles,\n",
    "                             alt_alleles,\n",
    "                             RESOLUTION\n",
    "                             )"
   ],
   "id": "d0e6eb38a4923166"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "plot_genotype_umi_comparison(gapfill_sdata, 'SET2', 'HEL',\n",
    "                             annotated_genotypes,\n",
    "                             celltype_genotypes,\n",
    "                             wt_alleles,\n",
    "                             alt_alleles,\n",
    "                             RESOLUTION\n",
    "                             )"
   ],
   "id": "693356df630d2173"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "pseudobulk_genotype_table(gapfill_sdata, 'JAK2 c.1849G>T', \"CAC\", \"AAC\", {\n",
    "    \"K562\": \"WT\",\n",
    "    \"HEL\": \"ALT\",\n",
    "    \"SET2\": \"HET\"\n",
    "})"
   ],
   "id": "689f886fc9df8f91"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "genotype_psuedobulk_accuracy_by_pcr(gapfill_sdata, 'JAK2 c.1849G>T', \"CAC\", \"AAC\", {\n",
    "    \"K562\": \"WT\",\n",
    "    \"HEL\": \"ALT\",\n",
    "    \"SET2\": \"HET\"\n",
    "}, ['K562', 'HEL'], max_threshold=25)"
   ],
   "id": "899d7776a5a78d9f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "for probe in gapfill_gdata.var.probe.unique():\n",
    "    if probe not in annotated_genotypes:\n",
    "        continue\n",
    "    wt_allele = wt_alleles[probe]\n",
    "    alt_allele = alt_alleles[probe]\n",
    "    ct_dict = {\n",
    "        \"HEL\": \"HET\" if len(celltype_genotypes[\"HEL\"][probe]) > 1 else (\"WT\" if wt_alleles[probe] in celltype_genotypes[\"HEL\"][probe] else \"ALT\"),\n",
    "        \"K562\": \"HET\" if len(celltype_genotypes[\"K562\"][probe]) > 1 else (\"WT\" if wt_alleles[probe] in celltype_genotypes[\"K562\"][probe] else \"ALT\"),\n",
    "        \"SET2\": \"HET\" if len(celltype_genotypes[\"SET2\"][probe]) > 1 else (\"WT\" if wt_alleles[probe] in celltype_genotypes[\"SET2\"][probe] else \"ALT\"),\n",
    "    }\n",
    "    genotype_psuedobulk_accuracy_by_pcr(gapfill_sdata, probe, wt_allele, alt_allele, ct_dict, [ct for ct, geno in ct_dict.items() if geno != \"HET\"], max_threshold=15)\n",
    "    print(probe)\n",
    "    print(\"---------------------------\")"
   ],
   "id": "234b453564683880"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from IPython.display import display\n",
    "for probe in gapfill_gdata.var.probe.unique():\n",
    "    probe_norm = probe.split(\"|\")\n",
    "    if len(probe_norm) > 1:\n",
    "        probe_norm = \" \".join(probe_norm[1:3])\n",
    "    else:\n",
    "        probe_norm = probe\n",
    "    if probe_norm not in annotated_genotypes:\n",
    "        continue\n",
    "    wt_allele = wt_alleles[probe_norm]\n",
    "    alt_allele = alt_alleles[probe_norm]\n",
    "    ct_dict = {\n",
    "        \"HEL\": \"HET\" if len(celltype_genotypes[\"HEL\"][probe_norm]) > 1 else (\"WT\" if wt_alleles[probe_norm] in celltype_genotypes[\"HEL\"][probe_norm] else \"ALT\"),\n",
    "        \"K562\": \"HET\" if len(celltype_genotypes[\"K562\"][probe_norm]) > 1 else (\"WT\" if wt_alleles[probe_norm] in celltype_genotypes[\"K562\"][probe_norm] else \"ALT\"),\n",
    "        \"SET2\": \"HET\" if len(celltype_genotypes[\"SET2\"][probe_norm]) > 1 else (\"WT\" if wt_alleles[probe_norm] in celltype_genotypes[\"SET2\"][probe_norm] else \"ALT\"),\n",
    "    }\n",
    "    # try:\n",
    "    df = pseudobulk_genotype_table(gapfill_sdata, probe, wt_allele, alt_allele, ct_dict)\n",
    "    display(df)\n",
    "    # except: continue\n",
    "    print(probe)\n",
    "    print(\"---------------------------\")"
   ],
   "id": "cc965e98e483bba1"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "genotype_cell_line_barplots(gapfill_sdata, 'JAK2 c.1849G>T', \"CAC\", \"AAC\", RESOLUTION)",
   "id": "1150754fbb8d15f8"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "genotype_accuracy_barplot(gapfill_sdata, 'JAK2 c.1849G>T', \"CAC\", \"AAC\", {\n",
    "    \"K562\": \"WT\",\n",
    "    \"HEL\": \"ALT\",\n",
    "    \"SET2\": \"HET\"\n",
    "}, RESOLUTION)"
   ],
   "id": "e2620b42321044e5"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "genotype_accuracy_barplot(gapfill_sdata, 'JAK2 c.1849G>T', \"CAC\", \"AAC\", {\n",
    "    \"K562\": \"WT\",\n",
    "    \"HEL\": \"ALT\",\n",
    "    \"SET2\": \"HET\"\n",
    "}, RESOLUTION, filter_na=False)"
   ],
   "id": "aef6f8c26f4ae3f9"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "for probe in gapfill_gdata.var.probe.unique():\n",
    "    if probe not in annotated_genotypes:\n",
    "        continue\n",
    "    wt_allele = wt_alleles[probe]\n",
    "    alt_allele = alt_alleles[probe]\n",
    "    ct_dict = {\n",
    "        \"HEL\": \"HET\" if len(celltype_genotypes[\"HEL\"][probe]) > 1 else (\"WT\" if wt_alleles[probe] in celltype_genotypes[\"HEL\"][probe] else \"ALT\"),\n",
    "        \"K562\": \"HET\" if len(celltype_genotypes[\"K562\"][probe]) > 1 else (\"WT\" if wt_alleles[probe] in celltype_genotypes[\"K562\"][probe] else \"ALT\"),\n",
    "        \"SET2\": \"HET\" if len(celltype_genotypes[\"SET2\"][probe]) > 1 else (\"WT\" if wt_alleles[probe] in celltype_genotypes[\"SET2\"][probe] else \"ALT\"),\n",
    "    }\n",
    "    fig, axes = plt.subplots(1, 2, figsize=(8, 4))\n",
    "    genotype_cell_line_barplots(gapfill_sdata, probe, wt_allele, alt_allele, RESOLUTION, ax=axes[0])\n",
    "    genotype_accuracy_barplot(gapfill_sdata, probe, wt_allele, alt_allele, ct_dict, RESOLUTION, filter_na=True, ax=axes[1])\n",
    "    plt.suptitle(probe)\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "    plt.clf()"
   ],
   "id": "b9b6fe8f31a939d7"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Now some basic visualization and analysis",
   "id": "fd073b46018166b2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Sort probes by capture rate and plot each on the spatial coords\n",
    "probes = gapfill_gdata.var.probe[~gapfill_gdata.var.probe.str.contains(\"0bp\")].unique()\n",
    "probes = sorted(probes, key=lambda x: gapfill_gdata[:, gapfill_gdata.var.probe == x].X.sum(), reverse=True)\n",
    "\n",
    "for probe in probes:\n",
    "    print(f\"Plotting {probe}\")\n",
    "    plot_genotypes(gapfill_sdata, probe, resolution=RESOLUTION)\n",
    "    plt.show()\n",
    "    plt.clf()"
   ],
   "id": "78053cbb3d052f73"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Cluster\n",
    "gapfill_sdata = gw.sp.recipe_spatial_expression_coclustering(\n",
    "    gapfill_sdata,\n",
    "    table_name=f\"square_{RESOLUTION:03d}um\",\n",
    "    n_highly_variable_genes=1000,\n",
    "    coordinate_system=\"\"\n",
    ")\n",
    "gapfill_sdata"
   ],
   "id": "9f16dddedada04ab"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Impute genotypes\n",
    "gapfill_sdata = gw.sp.impute_genotypes(\n",
    "    gapfill_sdata,\n",
    "    cluster_key='spatio_expression_coclustering',\n",
    "    resolution=f'square_{RESOLUTION:03d}um',\n",
    "    hold_out=0.05\n",
    ")\n",
    "gapfill_sdata"
   ],
   "id": "d6ea8d7df09d668f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Plot imputed genotypes\n",
    "for probe in probes:\n",
    "    print(f\"Plotting {probe}\")\n",
    "    plot_genotypes(gapfill_sdata, probe, resolution=RESOLUTION, imputed=True)\n",
    "    plt.show()\n",
    "    plt.clf()"
   ],
   "id": "b5903b1684e0219f"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "",
   "id": "29482deb1b432dee"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "boxplot_of_dualprobe_vs_gapfill(\n",
    "    dual_probe_sdata,\n",
    "    gapfill_sdata,\n",
    "    annotated_genotypes,\n",
    "    celltype_genotypes,\n",
    "    wt_alleles,\n",
    "    alt_alleles,\n",
    "    RESOLUTION\n",
    ")"
   ],
   "id": "38d6a555cee091a0"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
