---
title: GIFTwrap
summary: Tools for GIFT-seq data analysis
---
GIFTwrap
========
[![python](https://img.shields.io/badge/Python-3.11-3776AB.svg?style=flat&logo=python&logoColor=white)](https://www.python.org) [![PyPI](https://img.shields.io/pypi/v/giftwrap-sc.svg)](https://pypi.org/project/giftwrap-sc/) 

This package provides tools for dealing with GIFT-seq data. The `giftwrap` package is designed for both processing of GIFT-seq raw data and provides tools for downstream analysis of GIFT-seq data.

## Main Features
- **Data Processing**: Tools for processing GIFT-seq-based FastQ files into counts matrices.
- **Basic QC**: Automatic quality control of GIFT-seq data based on 10X Cell Ranger cell calls on the paired transcriptomes.
- **Downstream Analysis**: Functions for filtering and analyzing GIFT-seq data, including single-cell genotyping and imputation.
- **Spatial Analysis**: Tools for spatial analysis of GIFT-seq data, including visualization and spatial clustering.

## Installation
Please refer to our [installation guide](installation.md) for detailed instructions on how to install the `giftwrap` package.

## Usage
Please review our tutorials and examples to get started with using the `giftwrap` package for your GIFT-seq data analysis.

- [Getting Started](tutorials/getting_started.md)
- [Processing GIFT-seq Data](tutorials/processing_giftseq_data.md)
- [Analyzing GIFT-seq Data](tutorials/analyzing_giftseq_data.ipynb)
- [Imputing Genotypes](tutorials/imputation.md)
- [Spatial Analysis](tutorials/spatial_giftseq.md)
- [Seurat Integration](tutorials/seurat_integration.md)
- [Extending GIFTwrap](tutorials/extending_giftwrap.md)

## Contributing
We welcome contributions to the `giftwrap` package! Please see our [contributing guide](contributing.md) for more information on how to contribute to the project.

## Additional Information
For more information on GIFT-seq, please refer to our publication: ...
